README.md
# pyRM
Package to fit classical Rossiter-McLaughlin effect using MCMC.


Simple usage example:

```
python ./rm_fit.py --input=data/HD189733_t2-K2mask-filtered_ccf.rdb 
--exoplanet_priors=priors/HD189733.pars --calib_order=1 --nsteps=400 
--walkers=30 --burnin=150 --samples_filename=hd189733_mcmc_samples.h5 -v -k -r --output=HD_189733/ --basename= HD189733
```

The following input options are available:
```
--input to input data pattern (e.g., --input=*rdb.fits).
--exoplanet_priors to input file containing exoplanet priors (e.g., --exoplanet_priors=HD189733.pars)
--calib_priors to input file containing calibration priors (e.g., --calib_priors=calib.pars)
--calib_order to input the order of the polynomial that is added to the model for the calibration of each input dataset
--nsteps to input the number of steps in the MCMC analysis
--walkers to input the number of walkers in the MCMC analysis
--burnin to input the number of burn-in steps in the MCMC analysis
--samples_filename for the file name to save the MCMC samples
-v for verbose
-p for plotting
-k for saving the plots
-r for saving the detailed output in a csv file
--output for choosing the save directory of the plots and residuals
--basename for giving a stem to the output files common to all the datasets
```


changes.txt
Major additions before 28/04/2021 :

- added the "Normal_positive" prior type

- added the display of the rediduals parameters (overall, for each dataset, inside and outside of the transit) in the output

- added the display of the chi^2 in the output

- added two options :
	"-r" : saves the residuals for the datasets in a text file
	"-k" : keep the plots in the folder as png files. One can now choose if he wants to display the plots (using "-p"), save them, or both.
- Outputs the mean and standard deviation parameters for all the residuals. Useful when working with multiple datasets.
- Outputs the standard deviation of the residuals during, and outside of the transit, for each dataset.

- added the "--output" option that allows to choose the directory where the plots and residuals files are saved. By default it is the current directory.

- fixed the plots that did not display

Changes on 28/04/2021:

The calibration priors now include a center time. the associated prior in the priors file is "dnntc", where "nn" is the number of the dataset. This time has to be fixed to a constant. If no calibration prior is given, the center time is computed from the planet priors.

Output files have changed. Now we have one csv file for each dadaset with (for each point) the input data, the model, the keplerian, RM and Calibration parts, and the resiual. The option to get this output file is still "-r".

- added one option : "--basename" : gives a root basename for the files that imply all the datasets : global RV plot, corner plot, residuals histogram

- files related to only one dataset (individual RV plot and csv output) now take for base the name of the corresponding input file (only taking alphanumerical characters, dashes and underscores to avoid issues in file naming).

Changes on 21/05/2021:

- Added a new variable : phi0 . It is the time of periastron for the keplerian orbit. In a priors file, it has to be included just below tau. Before this version, the time of periastron was equal to the time for the center of transit, which caused issues for adjustment because of a bad positionning of the keplerian.

Changes on 27/05/2021 :

Adds two options for the alldatasets plot : 
--timerange that adds enlarge the time scale on each side of the plot by the value entered (in days)
--res that modifies the resolution of the plot model (but not the values of the adjustment!!). It can be set to very high (vh), high (h), medium (m, which is the default value), low (l), very low (vl) and extremely low (el). This parameter allows to earn computation time when plotting by changing the time step of the plotting (in order : 1/2 sec, 1 sec, 30 sec, 2 mn, 5 mn, 30 mn).

Added an option for all the plots : --format that allows to set the format of the plots. By default it is png, but of course it can be set to pdf, jpg etc.

Outputs a text file with the output planet and calibration parameters, the dispersion for each dataset and the chi^2
Fixed a bug that caused issues for the output data file if dealing with multiple datasets
Fixed some functions to avoid getting errors in the computation of the dispersions

priorslib.py
# -*- coding: iso-8859-1 -*-
"""
    Created on February 3 2020
    
    Description: Priors library
    
    @author: Eder Martioli <martioli@iap.fr>, Shweta Dalal <dalal@iap.fr>
    
    Institut d'Astrophysique de Paris, France.
    
    """

__version__ = "1.0"

__copyright__ = """
    Copyright (c) ...  All rights reserved.
    """

import numpy as np

def get_quantiles(dist, alpha = 0.68, method = 'median'):
    """
    get_quantiles function
    DESCRIPTION
        This function returns, in the default case, the parameter median and the error%
        credibility around it. This assumes you give a non-ordered
        distribution of parameters.
    OUTPUTS
        Median of the parameter,upper credibility bound, lower credibility bound
    """
    ordered_dist = dist[np.argsort(dist)]
    param = 0.0
    # Define the number of samples from posterior
    nsamples = len(dist)
    nsamples_at_each_side = int(nsamples*(alpha/2.)+1)
    if(method == 'median'):
       med_idx = 0
       if(nsamples%2 == 0.0): # Number of points is even
          med_idx_up = int(nsamples/2.)+1
          med_idx_down = med_idx_up-1
          param = (ordered_dist[med_idx_up]+ordered_dist[med_idx_down])/2.
          return param,ordered_dist[med_idx_up+nsamples_at_each_side],\
                 ordered_dist[med_idx_down-nsamples_at_each_side]
       else:
          med_idx = int(nsamples/2.)
          param = ordered_dist[med_idx]
          return param,ordered_dist[med_idx+nsamples_at_each_side],\
                 ordered_dist[med_idx-nsamples_at_each_side]

class normal_parameter:
      """
      Description
      -----------
      This class defines a parameter object which has a normal prior. It serves
      to save both the prior and the posterior chains for an easier check of the parameter.
      """
      
      def __init__(self,prior_hypp, pos = False):
          self.value = prior_hypp[0]
          self.value_u = 0.0
          self.value_l = 0.0
          self.prior_hypp = prior_hypp
          self.posterior = []
          self.is_positive = pos

      def get_ln_prior(self):
          return np.log(1./np.sqrt(2.*np.pi*(self.prior_hypp[1]**2)))-\
                 0.5*(((self.prior_hypp[0]-self.value)**2/(self.prior_hypp[1]**2)))

      def set_value(self,new_val):
          self.value = new_val

      def set_posterior(self,posterior_chain):
          self.posterior = posterior_chain
          param, param_u, param_l = get_quantiles(posterior_chain)
          self.value = param
          self.value_u = param_u
          self.value_l = param_l
      
      def check_value(self, x):
          return ((self.is_positive and x < 0) == False)

class uniform_parameter:
      """
      Description
      -----------
      This class defines a parameter object which has a uniform prior. It serves
      to save both the prior and the posterior chains for an easier check of the parameter.
      """
      def __init__(self,prior_hypp):
          self.value = (prior_hypp[0]+prior_hypp[1])/2.
          self.value_u = 0.0
          self.value_l = 0.0
          self.prior_hypp = prior_hypp
          self.posterior = []

      def get_ln_prior(self):
          return np.log(1./(self.prior_hypp[1]-self.prior_hypp[0]))

      def check_value(self,x):
          if x > self.prior_hypp[0] and  x < self.prior_hypp[1]:
              return True
          else:
              return False
 
      def set_value(self,new_val):
          self.value = new_val

      def set_posterior(self,posterior_chain):
          self.posterior = posterior_chain
          param, param_u, param_l = get_quantiles(posterior_chain)
          self.value = param
          self.value_u = param_u
          self.value_l = param_l

class jeffreys_parameter:
      """
      Description
      -----------
      This class defines a parameter object which has a Jeffreys prior. It serves
      to save both the prior and the posterior chains for an easier check of the parameter.
      """
      def __init__(self,prior_hypp):
          self.value = np.sqrt(prior_hypp[0]*prior_hypp[1])
          self.value_u = 0.0
          self.value_l = 0.0
          self.prior_hypp = prior_hypp
          self.posterior = []

      def get_ln_prior(self):
          return np.log(1.0) - np.log(self.value*np.log(self.prior_hypp[1]/self.prior_hypp[0]))

      def check_value(self,x):
          if x > self.prior_hypp[0] and  x < self.prior_hypp[1]:
              return True
          else:
              return False

      def set_value(self,new_val):
          self.value = new_val

      def set_posterior(self,posterior_chain):
          self.posterior = posterior_chain
          param, param_u, param_l = get_quantiles(posterior_chain)
          self.value = param
          self.value_u = param_u
          self.value_l = param_l

class constant_parameter:
      """
      Description
      -----------
      This class defines a parameter object which has a constant value. It serves
      to save both the prior and the posterior chains for an easier check of the parameter.
      """
      def __init__(self,val):
          self.value = val

def read_priors(filename, calibration = False):
    def generate_parameter(values):
        out_dict = {}
        out_dict['type'] = values[1]
        if values[1] == 'Normal':
            out_dict['object'] = normal_parameter(np.array(values[2].split(',')).astype('float64'))
        elif values[1] == 'Normal_positive':
            out_dict['object'] = normal_parameter(np.array(values[2].split(',')).astype('float64'), True)
        elif values[1] == 'Uniform':
            out_dict['object'] = uniform_parameter(np.array(values[2].split(',')).astype('float64'))
        elif values[1] == 'Jeffreys':
            out_dict['object'] = jeffreys_parameter(np.array(values[2].split(',')).astype('float64'))
        elif values[1] == 'FIXED':
            out_dict['object'] = constant_parameter(np.array(values[2].split(',')).astype('float64')[0])
        if len(values)>=5:
            out_dict['object'].set_value(np.float(values[3]))
        return out_dict
    # Open the file containing the priors:
    f = open(filename)
    # Generate dictionary that will save the data on the priors:
    priors = {}
    n_coefs = 0
    while True:
        line = f.readline()
        if line == '':
            break
        elif line[0] != '#':
            # Extract values from text file: [0]: parameter name,
            #                                [1]: prior type,
            #                                [2]: hyperparameters,
            #                                [3]: starting value (optional)
            values = line.split()
            priors[values[0]] = generate_parameter(values)
            errors = np.array(values[2].split(',')).astype('float64')
            error_key = "{0}_err".format(values[0])
            pdf_key = "{0}_pdf".format(values[0])
            priors[pdf_key] = values[1]
            priors[error_key] = errors
            n_coefs += 1

    f.close()

    if calibration :
        n_coefs-= 1
        ndatasets = n_coefs // (1+float(priors['orderOfPolynomial']['object'].value))
        priors["ndatasets"] = generate_parameter(['ndatasets','FIXED',str(ndatasets)])
    return priors

def read_exoplanet_params(prior_dict, output_theta_params = False):

    param_ids = ['per','tau','phi0','k','omega','ecc','rv0','lambda', 'vsini','a_R','inc','r_R','omega_rm','ldc']

    planet_params = {}
    
    for i in range(len(param_ids)):
        param = prior_dict[param_ids[i]]
        planet_params[param_ids[i]] = param['object'].value
    
    if output_theta_params :
        theta, labels = [], []
        for key in planet_params.keys():
            param = prior_dict[key]
            if param['type'] != 'FIXED':
                theta.append(planet_params[key])
                labels.append(key)
        return theta, labels
    else :
        for i in range(len(param_ids)):
            error_key = "{0}_err".format(param_ids[i])
            planet_params[error_key] = prior_dict[error_key]
            pdf_key = "{0}_pdf".format(param_ids[i])
            planet_params[pdf_key] = prior_dict[pdf_key]
        return planet_params

def read_calib_params(calib_priors_dict, output_theta_params = False):

    order_of_polynomial = calib_priors_dict['orderOfPolynomial']['object'].value
    ndatasets = calib_priors_dict['ndatasets']['object'].value

    calib_params = {}
    
    for i in range(int(ndatasets)):
        to_id = 'd{0:02d}tc'.format(i)
        calib_params[to_id] = calib_priors_dict[to_id]['object'].value
        for c in range(int(order_of_polynomial)):
            coeff_id = 'd{0:02d}c{1:1d}'.format(i,c)
            calib_params[coeff_id] = calib_priors_dict[coeff_id]['object'].value

    if output_theta_params :
        theta, labels = [], []
        for key in calib_params.keys():
            param = calib_priors_dict[key]
            if param['type'] != 'FIXED':
                theta.append(calib_params[key])
                labels.append(key)
        return theta, labels
    else :
        return calib_params

#intialize calib parameters
def init_calib_priors(tc, ndim=1, order=1, coefs=None) :
    """
        ndim : number of polynomials, typically the number of different datasets
        order:  number of coefficients in the polynomial
        coefs: array of arrays of coefficients.
        
        e.g.: if ndim=2, then coefs=[[3,2],[1,3]]
        """
    priors = {}

    order_dict = {}
    order_dict['type'] = "FIXED"
    order_dict['object'] = constant_parameter(order)
    priors['orderOfPolynomial'] = order_dict
    
    nds_dict = {}
    nds_dict['type'] = "FIXED"
    nds_dict['object'] = constant_parameter(ndim)
    priors['ndatasets'] = nds_dict
    
    for i in range(int(ndim)):
        to_id = 'd{0:02d}tc'.format(i)
        t_dict = {}
        t_dict['type'] = "FIXED"
        t_dict['object'] = constant_parameter(tc)
        priors[to_id] = t_dict
        for c in range(order):
            coeff_id = 'd{0:02d}c{1:1d}'.format(i,c)
            out_dict = {}
            out_dict['type'] = "Uniform"
            l_value = -1e20
            u_value = +1e20
            
            out_dict['object'] = uniform_parameter(np.array([l_value, u_value]))
            
            if coefs != None :
                out_dict['object'].set_value(np.float(coefs[i][c]))

            priors[coeff_id] = out_dict

    return priors


def get_theta_from_priors(planet_priors, calib_priors) :
    
    theta_planet, planet_labels = read_exoplanet_params(planet_priors, output_theta_params = True)

    theta_calib, calib_labels = read_calib_params(calib_priors, output_theta_params = True)

    theta = np.concatenate((theta_planet, theta_calib), axis=0)

    labels = planet_labels + calib_labels

    theta_priors = {}
    for key in planet_priors.keys() :
        if key in labels :
            theta_priors[key] = planet_priors[key]
    for key in calib_priors.keys() :
        if key in labels :
            theta_priors[key] = calib_priors[key]

    return theta, labels, theta_priors


def save_posterior(output, params, theta_fit, theta_labels, theta_err, calib=False, ncoeff=1, ref_time=0.) :

    outfile = open(output,"w")
    outfile.write("# Parameter_ID\tPrior_Type\tValues\n")
    if calib :
        outfile.write("orderOfPolynomial\tFIXED\t{0}\n".format(ncoeff))
    
    for key in params.keys() :
        if key in theta_labels :
            idx = theta_labels.index(key)
            error = (theta_err[idx][0] + theta_err[idx][1]) / 2.
            if key == 'tau' :
                outfile.write("{0}\tNormal\t{1:.4f},{2:.4f}\n".format(key, theta_fit[idx]+ref_time, error))
            else :
                outfile.write("{0}\tNormal\t{1:.4f},{2:.4f}\n".format(key, theta_fit[idx], error))
        else :
            if key == 'tau' :
                outfile.write("{0}\tFIXED\t{1}\n".format(key,params[key]+ref_time))
            else :
                outfile.write("{0}\tFIXED\t{1}\n".format(key,params[key]))

    outfile.close()


rm_fit.py
# -*- coding: iso-8859-1 -*-
"""
    Created on January 31 2020
    
    Description: Fit Rossiter-McLaughlin to RV data
    
    @author: Eder Martioli <martioli@iap.fr>, Shweta Dalal <dalal@iap.fr>
    
    Institut d'Astrophysique de Paris, France.
    
    Simple usage examples:
    
    python ./rm_fit.py --input=data/HD189733*.rdb --exoplanet_priors=priors/HD189733.pars --calib_order=1  --nsteps=400 --walkers=30 --burnin=150 --samples_filename=hd189733_mcmc_samples.h5 -v

    """

__version__ = "1.0"

__copyright__ = """
    Copyright (c) ...  All rights reserved.
    """


from optparse import OptionParser
import os,sys

import numpy as np
import glob
import rm_lib
import priorslib
from rm_lib import lnprob
import matplotlib.pyplot as plt
from astropy.io import ascii
import emcee

parser = OptionParser()
parser.add_option("-i", "--input", dest="input", help="Input RV data files",type='string',default="*.rdb")
parser.add_option("-t", "--output", dest="output", help="output files",type='string',default="")
parser.add_option("-m", "--basename", dest="basename", help="basename for output files",type='string',default="")
parser.add_option("-e", "--exoplanet_priors", dest="exoplanet_priors", help="File containing exoplanet priors",type='string',default="")
parser.add_option("-c", "--calib_priors", dest="calib_priors", help="File containing calibration priors",type='string',default="")
parser.add_option("-o", "--calib_order", dest="calib_order", help="Order of calibration polynomial",type='string',default="1")
parser.add_option("-n", "--nsteps", dest="nsteps", help="Number of MCMC steps",type='string',default="300")
parser.add_option("-w", "--walkers", dest="walkers", help="Number of MCMC walkers",type='string',default="32")
parser.add_option("-b", "--burnin", dest="burnin", help="Number of MCMC burn-in samples",type='string',default="100")
parser.add_option("-s", "--samples_filename", dest="samples_filename", help="MCMC samples file name",type='string',default="")
parser.add_option("-p", action="store_true", dest="plot", help="plot", default=False)
parser.add_option("-v", action="store_true", dest="verbose", help="verbose", default=False)
parser.add_option("-k", action="store_true", dest="save_plots", help="save_plots", default=False)
parser.add_option("-r", action="store_true", dest="save_residuals", help="save_residuals", default=False)
parser.add_option("-g", "--timerange", dest="timerange", help="add time range for alldatasets plot (in days)",type='string',default="0")
parser.add_option("-f", "--format", dest="fmt", help="format of the outpur plots",type='string',default="png")
parser.add_option("-l", "--res", dest="res", help="alldataset plot resolution",type='string',default="m")

try:
    options,args = parser.parse_args(sys.argv[1:])
except:
    print("Error: check usage with rm_fit.py -h ")
    sys.exit(1)

if options.verbose:
    print('Input RV data files: ', options.input)
    print('Exoplanet priors file: ', options.exoplanet_priors)
    print('Calibration priors file: ', options.calib_priors)
    print('Order of calibration: ', options.calib_order)
    print('Number of MCMC steps: ', options.nsteps)
    print('Number of MCMC walkers: ', options.walkers)
    print('Number of MCMC burn-in samples: ', options.burnin)
    print('MCMC samples file name: ', options.samples_filename)
    print('output directory : ', options.output)

calib_order = int(options.calib_order)
od = options.output
inf = options.input
tr = float(options.timerange)
fmt = options.fmt
res = options.res

planet_posterior = (options.exoplanet_priors).replace(".pars", "_posterior.pars")
if options.verbose:
    print("Output PLANET posterior: ", planet_posterior)

if options.calib_priors != "":
    calib_posterior = (options.calib_priors).replace(".pars", "_posterior.pars")
else :
    calib_posterior = "calibration_posterior.pars"

if options.verbose:
    print("Output CALIBRATION posterior: ", calib_posterior)

# make list of tfits data files
if options.verbose:
    print("Creating list of RV time series files...")
inputdata = sorted(glob.glob(options.input))
if options.basename != "":
    bn = options.basename
else:
    bn = ""
    for char in inputdata[0].rsplit('/')[-1]:
        if char.isalnum():
            bn += char
        if char in ["_","-"]:
            bn += char
# Load data
bjd, rvs, rverrs  = [], [], []
for i in range(len(inputdata)) :
    data = ascii.read(inputdata[i],data_start=2)
    bjd.append(np.array(data['rjd'] + 2400000.))
    rvs.append(np.array(data["vrad"]))
    rverrs.append(np.array(data["svrad"]))

# Load priors:
# Load exoplanet parameters priors
planet_priors = priorslib.read_priors(options.exoplanet_priors)
planet_params = priorslib.read_exoplanet_params(planet_priors)


# print out planet priors
if options.verbose:
    print("----------------")
    print("Input PLANET parameters:")
    for key in planet_params.keys() :
        if ("_err" not in key) and ("_pdf" not in key) :
            pdf_key = "{0}_pdf".format(key)
            if planet_params[pdf_key] == "FIXED" :
                print("{0} = {1} ({2})".format(key, planet_params[key], planet_params[pdf_key]))
            elif planet_params[pdf_key] == "Uniform" or planet_params[pdf_key] == "Jeffreys":
                error_key = "{0}_err".format(key)
                min = planet_params[error_key][0]
                max = planet_params[error_key][1]
                print("{0} <= {1} <= {2} ({3})".format(min, key, max, planet_params[pdf_key]))
            elif planet_params[pdf_key] == "Normal" :
                error_key = "{0}_err".format(key)
                error = planet_params[error_key][1]
                print("{0} = {1} +- {2} ({3})".format(key, planet_params[key], error, planet_params[pdf_key]))
    print("----------------")

# if priors file is provided then load calibration parameters priors
if options.calib_priors != "" :
    calib_priors = priorslib.read_priors(options.calib_priors, calibration = True)
    calib_params = priorslib.read_calib_params(calib_priors)
else :
#if no priors file is provided then make a guess
    tc = planet_priors['tau']['object'].value
    calib_priors = priorslib.init_calib_priors(tc, ndim=len(inputdata), order=calib_order)
    calib_params = priorslib.read_calib_params(calib_priors)
    for i in range(len(rvs)) :
        coeff_name = 'd{0:02d}c0'.format(i)
        calib_params[coeff_name] = np.mean(rvs[i]) - np.mean(rm_lib.rv_model(planet_params, bjd[i]))
        calib_priors[coeff_name]['object'].value = np.mean(rvs[i]) - np.mean(rm_lib.rv_model(planet_params, bjd[i]))

# print out calibration priors
if options.verbose:
    print("----------------")
    print("Input CALIBRATION parameters:")
    for key in calib_params.keys() :
        print(key, "=", calib_params[key])
    print("----------------")

# Variable "theta" stores only the free parameters, and "labels" stores the corresponding parameter IDs
theta, labels, theta_priors = priorslib.get_theta_from_priors(planet_priors, calib_priors)

# Print out free parameters, i.e., theta parameters
if options.verbose:
    print("----------------")
    print("Free parameters:")
    for i in range(len(theta)) :
        print(labels[i], " -> initial guess:", theta[i])
    print("----------------")

"""
# Uncomment this part to plot model and data for the intial guess
for i in range(len(bjd)) :
    calib = rm_lib.calib_model(len(bjd), i, calib_params, bjd[i])
    rvcurve = rm_lib.rv_model(planet_params, bjd[i])
    plt.plot(bjd[i],rvcurve + calib, '-')
    plt.errorbar(bjd[i], rvs[i], yerr=rverrs[i], fmt='o')
plt.show()
exit()
"""

#- initialize emcee sampler
amp, ndim, nwalkers, niter, burnin = 5e-4, len(theta), int(options.walkers), int(options.nsteps), int(options.burnin)

# Set up the backend
if options.samples_filename != "":
    backend = emcee.backends.HDFBackend(od+options.samples_filename)
    backend.reset(nwalkers, ndim)
else :
    backend = None

sampler = emcee.EnsembleSampler(nwalkers, ndim, lnprob, args = [theta_priors, labels, calib_params, planet_params, bjd, rvs, rverrs], backend=backend)
pos = [theta + amp * np.random.randn(ndim) for i in range(nwalkers)]
#--------

#- run mcmc
if options.verbose:
    print("Running MCMC ...")
sampler.run_mcmc(pos, niter, progress=True)
samples = sampler.chain[:, burnin:, :].reshape((-1, ndim)) # burnin : number of first samples to be discard as burn-in
#--------

# Obtain best fit calibration parameters from pdfs
calib_params, calib_theta_fit, calib_theta_labels, calib_theta_err = rm_lib.best_fit_params(calib_params, labels, samples)

# Obtain best fit planet parameters from pdfs
planet_params, planet_theta_fit, planet_theta_labels, planet_theta_err = rm_lib.best_fit_params(planet_params, labels, samples)

p = options.plot
k = options.save_plots
r = options.save_residuals

#saves the output
if k:
    f = open(od+"Parameters output.txt",'w')
    f.truncate(0)

# print out best fit parameters and errors
if options.verbose:
    print("----------------")
    print("PLANET Fit parameters:")
    f.write("PLANET Fit parameters:\n")
    for i in range(len(planet_theta_fit)) :
        print(planet_theta_labels[i], "=", planet_theta_fit[i], "+", planet_theta_err[i][0], "-", planet_theta_err[i][1])
        f.write(str(planet_theta_labels[i])+" = " +str(planet_theta_fit[i])+" + "+str(planet_theta_err[i][0])+" - "+str(planet_theta_err[i][1])+"\n")

    print("----------------")
    f.write("----------------\n")
    print("CALIBRATION Fit parameters:")
    f.write("CALIBRATION Fit parameters:\n")
    for i in range(len(calib_theta_fit)) :
        print(calib_theta_labels[i], "=", calib_theta_fit[i], "+", calib_theta_err[i][0], "-", calib_theta_err[i][1])
        f.write(str(calib_theta_labels[i])+" = "+str(calib_theta_fit[i])+" + "+str(calib_theta_err[i][0])+" - "+str(calib_theta_err[i][1])+'\n')
    print("----------------")
    f.write("----------------\n")

bjd_limits=[]


if p or k :
    print("Plotting :")
    #plot all data sets
    rm_lib.plot_all_datasets(bjd, rvs, rverrs, planet_params, calib_params, samples, labels, theta_priors, tr, tr, fmt, res, bn, p, k, od)

    # plot each dataset with the best model
    for i in range(len(bjd)) :
        rm_lib.plot_individual_datasets(bjd, rvs, rverrs, i, planet_params, calib_params, samples, labels, theta_priors, fmt,inf, p, k, od, bjd_limits=bjd_limits, detach_calib=False)

# save posterior of planet parameters into file:
priorslib.save_posterior(planet_posterior, planet_params, planet_theta_fit, planet_theta_labels, planet_theta_err)

# save posterior of calibration parameters into file:
ncoeff=calib_priors['orderOfPolynomial']['object'].value
priorslib.save_posterior(calib_posterior, calib_params, calib_theta_fit, calib_theta_labels, calib_theta_err, calib=True, ncoeff=ncoeff)

if p or k :
    #- make a pairs plot from MCMC output:
    rm_lib.pairs_plot(samples, labels, calib_params, planet_params, fmt, bn, p, k, od)
    #--------

if p or k or r:
    #- perform analysis of residuals:
    print("----------------")
    print("Running Analysis of residuals:")
    f.write("Analysis of residuals :\n----------------\n")
    f.close()
    rm_lib.analysis_of_residuals(bjd, rvs, rverrs, planet_params, calib_params, theta_priors, bn, fmt, inf, p, k, r, od, output="")
    #--------
